_source/_posts/2020-12-11-build-and-secure-an-api-in-python-with-fastapi.md

```diff
@@ -0,0 +1,362 @@
+
+## What Sets FastAPI Apart?
+
+Unlike many web frameworks, FastAPI is built to support REST APIs first. As [single-page applications become increasingly common](https://w3-lab.com/web-development/single-page-applications-pros-2020/), separate backends like those created with FastAPI are an essential part of web application architecture. With this in mind, FastAPI incorporates several features meant to support HTTP API-driven application development.
+
+First, documentation is created for your application automatically. Using [OpenAPI](https://www.openapis.org/), FastAPI generates clean, interactive documentation for your application. Later in this tutorial, you'll see how you can use this documentation to test your API's authorization functionality.
+
+Next, FastAPI is, as its name suggests, [fast](https://fastapi.tiangolo.com/benchmarks/). Python has historically tended to be slower than Node or Go because of the blocking, synchronous nature of the language; in Python 3, the option to declare functions `async` changed that. Some older frameworks aren't built to support asynchronous Python, but FastAPI does. It's built on [Starlette](https://www.starlette.io/), a lightweight asynchronous Python server, and it takes advantage of modern features like [typing](https://docs.python.org/3/library/typing.html) to improve developer experience.
+
+Another feature implemented by FastAPI, but not necessarily common in the Python ecosystem, is [dependency injection](https://fastapi.tiangolo.com/features/#dependency-injection). The FastAPI framework uses type hints in your function signatures to find and inject the required dependencies. This allows you to write decoupled, reusable code, making unit testing much easier as dependencies can be swapped out at runtime.
+
+Finally, while FastAPI comes with many of the features you would expect in a REST API framework (like [data validation](https://fastapi.tiangolo.com/features/#validation) and [authentication](https://fastapi.tiangolo.com/features/#security-and-authentication)), it lets you choose your ORM and database of choice. This means that FastAPI can work with your existing data models if you're migrating from an existing Python application.
+
+While FastAPI is an excellent option for building REST APIs in Python, it's not perfect for every situation. If the system you're building relies on Python 3.5 or any earlier Python framework, you won't be able to use FastAPI. FastAPI also isn't a great choice if you're building a server-side web application that needs to display HTML, as it doesn't include all the view helper functions that full-stack frameworks like Django do.
+
+Now that you're familiar with FastAPI at a high level, you're ready to start building your first application. In the rest of this tutorial, you'll see how to get started and secure your endpoints using Okta as your OAuth authorization server.
+
```

Comment:
I like to do a prerequisites and table of contents right at the beginning. Here's a sample:

```
**Prerequisites**:

- [Node 14](https://nodejs.org/)
- [Okta CLI 0.7.1+](https://github.com/okta/okta-cli)

**Table of Contents**{: .hide }
* Table of Contents
{:toc}
```